language/move-prover/mutation/src/mutator.rs

```diff
@@ -0,0 +1,233 @@
+        );
+    let matches = cmd_line_parser.get_matches_from(args);
+    let get_vec = |s: &str| -> Vec<String> {
+        match matches.values_of(s) {
+            Some(vs) => vs.map(|v| v.to_string()).collect(),
+            _ => vec![],
+        }
+    };
+    let sources = get_vec("sources");
+    let deps = get_vec("dependencies");
+    let configs: Vec<Option<String>> = if matches.is_present("config") {
+        get_vec("config").into_iter().map(Some).collect_vec()
+    } else {
+        vec![None]
+    };
+
+    for config_spec in configs {
+        let (config, out) = if let Some(config_file) = &config_spec {
+            let extension = "mod_data";
+            let out = PathBuf::from(config_file)
+                .with_extension(extension)
+                .to_string_lossy()
+                .to_string();
+            (config_spec, out)
+        } else {
+            (None, "mutation.data".to_string())
+        };
+        if let Err(s) = apply_mutation(config.as_ref(), &sources, &deps) {
+            println!("ERROR: execution failed: {}", s);
+        } else {
+            println!("results stored at `{}`", out);
+        }
+    }
+}
+
+fn apply_mutation(
+    config_file_opt: Option<&String>,
+    modules: &[String],
+    dep_dirs: &[String],
+) -> anyhow::Result<()> {
+    println!("building model");
+    let env = run_model_builder(modules, dep_dirs)?;
+    let mut error_writer = StandardStream::stderr(ColorChoice::Auto);
+    let mut options = if let Some(config_file) = config_file_opt {
+        Options::create_from_toml_file(config_file)?
+    } else {
+        Options::default()
+    };
+
+    // Do not allow any mutation to run longer than 100 seconds to avoid extremely long use times
+    options.backend.hard_timeout_secs = 100;
+
+    options.verbosity_level = LevelFilter::Error;
+
+    options.prover.mutation = true;
+    options.backend.derive_options();
+    options.setup_logging();
+    check_errors(&env, &options, &mut error_writer, "unexpected build errors")?;
+
+    let config_descr = "default".to_string();
+
+    println!(
+        "Starting mutations with config `{}`.",
+        config_descr
+    );
+
+    let mut i = 0;
+    let mut mutation_applied = true;
+
+    let mut runner = Runner {
+        options,
+        error_writer,
+    };
+
+    while mutation_applied {
+        i += 1;
+        println!("Applying add-sub mutation {}", i);
+        runner.options.prover.mutation_add_sub = i;
+        env.set_extension(MutationManager {
+            mutated: false,
+            add_sub: i,
+        });
+        mutation_applied = runner.mutate(&env)?;
+    }
+    Ok(())
+}
+
+impl Runner {
+    fn mutate(&mut self, env: &GlobalEnv) -> anyhow::Result<bool> {
+        let mut mutated = false;
+        for module in env.get_modules() {
+            if module.is_target() {
+                self.mutate_module(module)?;
+                let mm = env.get_extension::<MutationManager>();
```

Comment:
nit: `env.get_extension::<MutationManager>().map(|e| e.mutated).unwrap_or(false)`

Reply:
Updated, thanks for the tip!